bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java

```diff
@@ -133,18 +171,33 @@ public void stop()
 
         List<Object> reversedInstances = new ArrayList<>(managedInstances);
         Collections.reverse(reversedInstances);
+        List<Throwable> failures = new ArrayList<>();
 
         for (Object obj : reversedInstances) {
             log.debug("Stopping %s", obj.getClass().getName());
             LifeCycleMethods methods = methodsMap.get(obj.getClass());
             for (Method preDestroy : methods.methodsFor(PreDestroy.class)) {
                 log.debug("\t%s()", preDestroy.getName());
-                preDestroy.invoke(obj);
+                try {
+                    preDestroy.invoke(obj);
+                }
+                catch (Exception e) {
+                    Throwable t = (e instanceof InvocationTargetException) ? e.getCause() : e;
+                    if (logFailuresImmediately) {
+                        log.error(t, "Exception in PreDestroy method %s::%s()", obj.getClass().getName(), preDestroy.getName());
+                    }
+                    else {
+                        log.debug(t, "Exception in PreDestroy method %s::%s()", obj.getClass().getName(), preDestroy.getName());
```

Comment:
We shouldn't log here at all. Even at debug level, log-and-throw is still confusing.